core/src/price_estimation/clients/oneinch/api.rs

```diff
@@ -0,0 +1,179 @@
+use super::super::generic_client::{Api, GenericToken};
+use crate::http::{HttpClient, HttpFactory, HttpLabel};
+use anyhow::{Context, Result};
+use ethcontract::Address;
+use futures::future::{BoxFuture, FutureExt as _};
+use serde::Deserialize;
+use std::collections::HashMap;
+use url::Url;
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    pub address: Option<Address>,
+    pub decimals: u8,
+}
+impl GenericToken for Token {
+    fn symbol(&self) -> &str {
+        &self.symbol
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TradedAmounts {
+    to_token_amount: String,
+    from_token_amount: String,
+}
+
+#[derive(Debug)]
+pub struct OneinchHttpApi {
+    api_url: Url,
+    client: HttpClient,
+}
+impl OneinchHttpApi {
+    pub fn with_url(http_factory: &HttpFactory, api_url: &str) -> Result<Self> {
+        let client = http_factory
+            .create()
+            .context("failed to initialize HTTP client")?;
+        let api_url = api_url
+            .parse()
+            .with_context(|| format!("failed to parse url {}", api_url))?;
+        Ok(OneinchHttpApi { api_url, client })
+    }
+}
+
+pub const DEFAULT_API_URL: &str = "https://api.1inch.exchange/v1.1/";
+
+/// 1inch API version v1.1
+/// https://1inch.exchange/#/api
+impl Api for OneinchHttpApi {
+    type Token = Token;
+
+    fn bind(http_factory: &HttpFactory) -> Result<Self> {
+        Self::with_url(http_factory, DEFAULT_API_URL)
+    }
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Token>>> {
+        async move {
+            let url = self.api_url.join("tokens").unwrap();
+            let token_mapping: HashMap<String, Token> = self
+                .client
+                .get_json_async(url.to_string(), HttpLabel::Oneinch)
+                .await
+                .context("failed to parse token list json from 1inch response")?;
+            Ok(token_mapping
+                .into_iter()
+                .map(|(_token_symbol, token_data)| token_data)
+                .collect())
+        }
+        .boxed()
+    }
+
+    fn get_price<'a>(&'a self, from: &Token, to: &Token) -> BoxFuture<'a, Result<f64>> {
+        let mut url = self.api_url.join("quote").unwrap();
```

Comment:
nit: I assume we also unwrap in dex.ag, but technically we could `?` here to avoid panics.

Reply:
Actually URL management is a bit different in 1inch, so we don't need to unwrap there.

I can't just use `?` because I need to return a `BoxFuture`, and I can't wrap everything inside `async move` because of lifetime issues with `from` and `to`. I could clone `from.symbol` and `to.symbol` and then move that, but it's more code and I think it's actually better if the program fails when failing to create an url when joining two compile-time constants.

Reply:
I would suggest moving the URL computation into the `async { ... }` block so that you can use `?`.

Reply:
Hmm, maybe you could do:
```
let mut url = self.api_url.join("quote").and_then(move |url| {
    // use `to` and `from` to compute the URL here...
});
```

And use `url?` inside the `async { ... }` block?

The reason I suggest this is that the API `with_url` can be created with a caller specified URL which might error when `join`-ing `"quote"`.

Reply:
It makes sense, and I like this solution!